### ISSUE TITLE
Possible false positive "Potential leak of memory" on C union

### ISSUE THREAD

**Issue by sherburne** [llvm/llvm-project#60246]
I'm running into a memory leak report when analyzing the following C code:

    #include <stdlib.h>
    #include <stdio.h>
    #include <string.h>

    struct s {
        union {
            const char * const str;
        };
    };

    int main() {
        char * str = calloc(5, sizeof(char));
        strcpy(str, "test");
    
        struct s foo = {
            .str = str
        };
    
        struct s * fah = calloc(1, sizeof(struct s));
        memcpy(fah, &foo, sizeof(struct s));
    
        free((void *) fah->str);
        free(fah);
    }

And the analyzer output:

     <source>:19:22: warning: Potential leak of memory pointed to by 'str' [unix.Malloc]
         struct s * fah = calloc(1, sizeof(struct s));

I think this is a false positive. This code is also running here: [https://godbolt.org/z/bz5vzPa8h](https://godbolt.org/z/bz5vzPa8h)

`clang -v` output:

    Apple clang version 14.0.0 (clang-1400.0.29.202)
    Target: x86_64-apple-darwin21.6.0
    Thread model: posix
    InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

Please let me know if you need more information, or have any suggestions. Thanks!

**Comment by haoNoQ** [comment]
Yes, that's a false positive, thanks for a very useful bug report!

Looks like a dead symbol collection bug. It cleans up the conjured symbol produced by the first calloc (note that it's the *first* calloc that's "leaking") even though it's still clearly accessible through the non-lazy compound value in the cleaned state. The "non-lazy" part is what makes this rare; the `{ .str = str }` initializer of the union caused us to construct one (most of our compound values are "lazy" and we're reasonably good at handling them). But definitely important to fix nonetheless.

<img width="457" alt="Screenshot 2023-01-24 at 12 51 18 PM" src="https://user-images.githubusercontent.com/2094875/214409467-91c7cb52-0613-4036-a3ff-ab6e6274dba1.png">

**Comment by sherburne** [comment]
Good! Glad my weird code was helpful. 😄

**Comment by tianxinghe** [comment]
> Yes, that's a false positive, thanks for a very useful bug report!
> 
> Looks like a dead symbol collection bug. It cleans up the conjured symbol produced by the first calloc (note that it's the _first_ calloc that's "leaking") even though it's still clearly accessible through the non-lazy compound value in the cleaned state. The "non-lazy" part is what makes this rare; the `{ .str = str }` initializer of the union caused us to construct one (most of our compound values are "lazy" and we're reasonably good at handling them). But definitely important to fix nonetheless.
> 
> <img alt="Screenshot 2023-01-24 at 12 51 18 PM" width="457" src="https://user-images.githubusercontent.com/2094875/214409467-91c7cb52-0613-4036-a3ff-ab6e6274dba1.png">

I would like to inquire about the current situation of Clang's handling of the union, as it seems to have missed some defects.
For example:
    union u
    {
        int First;
        int Second;
    };
    
    int main() {
      int data;
      data = 0;
      union u n;
      n.First = data;
      int result = (int)(100 / n.Second);
      return 0;
    }

It fails to detect the Dividing by Zero.
Thank you!
